keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,309 @@
+
+### Custom taint-manager
+
+Opting out from the default `TaintManger`, Kubernetes operators will be able to implement custom logic to handle `NoExecute` taint such as delegating the authority for the termination for some classes of workloads providing high-availability protection guarantees. They can also scale out the usage of `NoExcute` to more use-cases than `NotReady` nodes (e.g. unhealthy network, etc.).
+
+<p align="center"><image src="custom-taint-manager.png" width=90%></p>
+
+### Implementation
+
+A new `NodeLifeCycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.
+
+A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.  A new flag is added 
+
+`// NoExecuteTaintManagerConfiguration contains elements describing NoExecuteTaintManagerConfiguration`
+`type NoExecuteTaintManagerConfiguration struct {`
+`// enables the default NoExecuteTaintManager. MUST be synced with the`
+`// corresponding flag of the kube-apiserver.`
+`EnableNoExecuteTaintManager bool`
+`  ...`
+`}`
+
+The creation and startup of the default taint-manager is performed by `kube-controller-manager` . 
+
+`func startNoExecuteTaintManager(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {`
+`    if !controllerContext.ComponentConfig.NoExecuteTaintManager.EnableNoExecuteTaintManager {
+        return nil, false, nil
+    }`
+
+```
+taintManager, err = taintmanager.NewNoExecuteTaintManager(ctx, kubeClient, nc.podLister, nc.nodeLister, nc.getPodsAssignedToNode)
+        nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+            AddFunc: controllerutil.CreateAddNodeHandler(func(node *v1.Node) error {
+                nc.taintManager.NodeUpdated(nil, node)
+                return nil
+            }),
+            UpdateFunc: controllerutil.CreateUpdateNodeHandler(func(oldNode, newNode *v1.Node) error {
+                nc.taintManager.NodeUpdated(oldNode, newNode)
+                return nil
+            }),
+            DeleteFunc: controllerutil.CreateDeleteNodeHandler(func(node *v1.Node) error {
+                nc.taintManager.NodeUpdated(node, nil)
+                return nil
+            }),
+        })
+    }
+        if err != nil {
+        return nil, true, fmt.Errorf("failed to start the taint manager: %v", err)
+    }
+  }  
+```
+
+### Test Plan
+
+* Unit and integration tests
+    * Add nodes taints
+    * Act on `NoExecute` to delete running pods
+    * Respect taint tolerations 
+* Verify passing the existing E2E and conformance tests.
+    * Taint a node 
+    * Evict the correct set of running pods
+
+### Graduation Criteria
+
+#### Alpha
+
+* Support `kube-controller-manaager` flag `controllers=-taint-manager` to disable the default taint-manager.
```

Comment:
Is the `-` semantics really what we're going to implement? I'd assume disabling a default controller can be achieved by providing a full list of controller names.

If this is a goal of this KEP, I highly recommend to document it in the goals section.

Reply:
Yes, it's already listed in the goal section.

Reply:
I guess this is side effect of making taint-manager a separate controller. This is not the goal?

Reply:
By decoupling the taint-manager from nodelifecyclecontroller, we can just use existing controller-manager flag to disable the default taint-manager instead of using a separate flag. I agree it's not the main goal. I'll remove it.